mlflow/pyfunc/dbconnect_artifact_cache.py

```diff
@@ -0,0 +1,135 @@
+            _global_dbconnect_artifact_cache = DBConnectArtifactCache(spark)
+            cache_file = os.path.join(get_or_create_tmp_dir(), _CACHE_MAP_FILE_NAME)
+            if is_in_databricks_runtime() and os.path.exists(cache_file):
+                # In databricks runtime (shared cluster or Serverless), when you restart the notebook
+                # REPL by %restart_python or dbutils.library.restartPython(), the DBConnect session is
+                # still preserved. So in this case, we can reuse the cached artifact files.
+                # So that when adding artifact, the cache map is serialized to local disk file
+                # `db_connect_artifact_cache.json` and after REPL restarts, `DBConnectArtifactCache`
+                # restores the cache map by loading data from the file.
+                with open(cache_file, "r") as f:
+                    _global_dbconnect_artifact_cache._cache = json.load(f)
+        return _global_dbconnect_artifact_cache
+
+    def __init__(self, spark):
+        self._spark = spark
+        self._cache = {}
+
+    def __getstate__(self):
+        """
+        The `DBConnectArtifactCache` instance is created in Databricks Connect client side,
+        and it will be pickled to Databricks Connect UDF sandbox
+        (see `get_unpacked_artifact_dir` method), but Spark Connect client object is
+        not pickle-able, we need to skip this field.
+        """
+        state = self.__dict__.copy()
+        # Don't pickle `_spark`
+        del state["_spark"]
+        return state
+
+    def __setstate__(self, state):
+        self.__dict__.update(state)
+        self._spark = None
+
+    def has_cache_key(self, cache_key):
+        return cache_key in self._cache
+
+    def add_artifact_archive(self, cache_key, artifact_archive_path):
+        """
+        Add an artifact archive file to Databricks connect cache.
+        The archive file must be 'tar.gz' format.
+        You can only call this method in Databricks Connect client side.
+        """
+        if not artifact_archive_path.endswith(".tar.gz"):
+            raise RuntimeError(
+                "'add_artifact_archive' only supports archive file in 'tar.gz' format."
+            )
+
+        archive_file_name = os.path.basename(artifact_archive_path)
+        if cache_key not in self._cache:
+            self._spark.addArtifact(artifact_archive_path, archive=True)
+            self._cache[cache_key] = archive_file_name
+
+        if is_in_databricks_runtime():
+            with open(os.path.join(get_or_create_tmp_dir(), _CACHE_MAP_FILE_NAME), "w") as f:
+                json.dump(self._cache, f)
+
+    def get_unpacked_artifact_dir(self, cache_key):
+        """
+        Get unpacked artifact directory path, you can only call this method
+        inside Databricks Connect spark UDF sandbox.
+        """
+        if cache_key not in self._cache:
+            raise RuntimeError(f"You haven't uploaded artifact '{cache_key}'.")
```

Comment:
who uploads file? The user? or mlflow?

Reply:
`add_artifact_archive` method uploads artifact

Reply:
When writing an error message, avoid using "you," as it can sound accusatory. Instead, use neutral language that focuses on the issue. For example, say "A valid input is required" instead of "You must provide a valid input." This promotes a supportive tone and helps users understand how to resolve the issue without feeling blamed.